### PR TITLE
🎉 Source Zendesk Chat: migrate to `YamlDeclarativeSource (Low-code)`

### PR DESCRIPTION
**Pull request by bazarnov**
## What
Resolving: https://github.com/airbytehq/airbyte-internal-issues/issues/6468

## How
- by using the airbyte-cdk.declarative (Low-code)
- added missing `pk` for `agents_timeline` stream: `id` to the schema
- added 5 custom components to cover the migration:
   - `bans_record_extractor` - custom `record_extractor` for `bans` stream
   - `id_incremental_cursor` - custom `cursor` to provide the ability to pull data using `id`(int) as cursor.
   - `id_offset_pagination_strategy` - the pagination strategy that uses the `id` offset from the `records` and uses it as `next_page_token`
   - `time_offset_pagination_strategy` - the pagination strategy that uses `timestamp` from `response` as the `next_page_token`
   - `timestamp_based_cursor` - provides the ability to track the `cursor_field` values from the `record` instead of `stream_slices` and making adjustments to the `request_params` needed for `agent_timeline` and `chats` streams
- covered all custom components with `unit_tests`
- removed `leftovers` from the previous code-base
- added `migrations.md` to pass `QA` checks as of `breaking change`
- updated `expected records` to match the changes

## 🚨 User Impact 🚨
The breaking changes are avoided, with additional RECORD Transformations.

### CODE REVIEW

File: airbyte-integrations/connectors/source-zendesk-chat/source_zendesk_chat/manifest.yaml
Comment (maxi297) [inline airbyte-integrations/connectors/source-zendesk-chat/source_zendesk_chat/manifest.yaml:267]: Differences observed:
* Query param `start_time` changes from `1601510400000000` to `1601524800000000` (this will probably give you another value depending on your timezone). So I've learned today that parsing to `%s` format ignores the timezone so that `datetime.datetime.fromisoformat("2020-10-01T00:00:00+00:00").strftime("%s")` will return an epoch assuming `2020-10-01T00:00:00` was in your timezone and the `+00:00`. Based on the [IBM documentation](https://www.ibm.com/docs/en/i/7.3?topic=functions-strftime-convert-datetime-string#strfti__title__5), this behavior can be affected by the system timezone. So I guess there is a bug in the CDK which we should fix by adding [this if](https://github.com/airbytehq/airbyte/blob/0113c525548080f128258042d8663a415897ef8c/airbyte-cdk/python/airbyte_cdk/sources/declarative/datetime/datetime_parser.py#L28-L29) [here](https://github.com/airbytehq/airbyte/blob/0113c525548080f128258042d8663a415897ef8c/airbyte-cdk/python/airbyte_cdk/sources/declarative/interpolation/macros.py#L108-L117). I have created [an issue](https://github.com/airbytehq/airbyte-internal-issues/issues/6545) for that
* The id changes from `360786799676|2020-12-14T04:08:32Z` to `360786799676|2020-12-14T04:08:32.301292Z`. I'm not sure why as [the code from the previous version](https://github.com/airbytehq/airbyte/blob/0113c525548080f128258042d8663a415897ef8c/airbyte-integrations/connectors/source-zendesk-chat/source_zendesk_chat/streams.py#L210) seems very similar. Can you investigate on that one?

File: airbyte-integrations/connectors/source-zendesk-chat/source_zendesk_chat/components/id_incremental_cursor.py
Comment (maxi297) [inline airbyte-integrations/connectors/source-zendesk-chat/source_zendesk_chat/components/id_incremental_cursor.py:82]: Where/when is this called? Semi-incremental is not yet part of the CDK and hence, I don't think this does anything

**Review by maxi297 — COMMENTED** [review]
Thanks @bazarnov  ! Can you acknowledge that the change in behaviors that were identified are fine?

File: airbyte-integrations/connectors/source-zendesk-chat/source_zendesk_chat/manifest.yaml
Comment (bazarnov) [reply airbyte-integrations/connectors/source-zendesk-chat/source_zendesk_chat/manifest.yaml:85]: Based on this doc: https://developer.zendesk.com/api-reference/live-chat/chat-api/routing_settings/#update-account-routing-settings
the it's not very promising to have for the rest of the streams, except of `routing_settings`, but even more, it's not applicable for GET (the data fetch), but for the PUT (updating the agent's settings from the API). Whatever, I'll get rid of it completely.

File: airbyte-integrations/connectors/source-zendesk-chat/source_zendesk_chat/manifest.yaml
Comment (bazarnov) [reply airbyte-integrations/connectors/source-zendesk-chat/source_zendesk_chat/manifest.yaml:175]: The `Accounts` stream doesn't have the `Pagination` at all, returning the List[dict] in response. Yeah, because of the same `base_stream` it would apply this request param, which is fine to pass across the diff endpoints we have for this source.

File: airbyte-integrations/connectors/source-zendesk-chat/source_zendesk_chat/manifest.yaml
Comment (bazarnov) [reply airbyte-integrations/connectors/source-zendesk-chat/source_zendesk_chat/manifest.yaml:195]: Corrected.

File: airbyte-integrations/connectors/source-zendesk-chat/source_zendesk_chat/manifest.yaml
Comment (bazarnov) [reply airbyte-integrations/connectors/source-zendesk-chat/source_zendesk_chat/manifest.yaml:252]: Thanks for highlighting it!

File: airbyte-integrations/connectors/source-zendesk-chat/source_zendesk_chat/components/bans_record_extractor.py
Comment (bazarnov) [reply airbyte-integrations/connectors/source-zendesk-chat/source_zendesk_chat/components/bans_record_extractor.py:26]: After we merge 2 pieces `ip_address` and `visitor`, which have basically the same structure, but resolved into 2 properties for the single response object: `{ "ip_address": {...}, "visitor": {...} }`, we sort the records additionally by the created at, to prevent the STATE collisions later on. But originally, this was added 3 years ago, and didn't change by now. Decided to leave it as it, instead of thinking what it the best approach here to solve the unsorted records. If we do have the same functionality inside the Low-Code, we can try to optimize this part for this source.

File: airbyte-integrations/connectors/source-zendesk-chat/source_zendesk_chat/components/bans_record_extractor.py
Comment (bazarnov) [reply airbyte-integrations/connectors/source-zendesk-chat/source_zendesk_chat/components/bans_record_extractor.py:27]: Good catch! Corrected.

File: airbyte-integrations/connectors/source-zendesk-chat/source_zendesk_chat/components/id_incremental_cursor.py
Comment (bazarnov) [reply airbyte-integrations/connectors/source-zendesk-chat/source_zendesk_chat/components/id_incremental_cursor.py:82]: This is the `abstractmethod` set in the super-class: 
https://github.com/airbytehq/airbyte/blob/0113c525548080f128258042d8663a415897ef8c/airbyte-cdk/python/airbyte_cdk/sources/declarative/incremental/cursor.py#L55

It has to be defined, even if it does nothing.

File: airbyte-integrations/connectors/source-zendesk-chat/source_zendesk_chat/components/bans_record_extractor.py
Comment (maxi297) [reply airbyte-integrations/connectors/source-zendesk-chat/source_zendesk_chat/components/bans_record_extractor.py:26]: > to prevent the STATE collisions later on

I'm slow this morning: Can you clarify this point? Which situation would generate state collision and what would be the impact?

File: airbyte-integrations/connectors/source-zendesk-chat/source_zendesk_chat/manifest.yaml
Comment (maxi297) [reply airbyte-integrations/connectors/source-zendesk-chat/source_zendesk_chat/manifest.yaml:269]: Given my understanding, it could lead to duplication of records if we do not fix this and don't ask the user to reset the stream. The case would be:
* First sync on non low-code version, state cursor value is the one of the last record
* Source is updated
* Sync a second time from the last record (so last record will be picked up again) but ID created is different from the first sync so dedup is not happening

Can you try implementing the fix? I'm up-to-date on `origin/baz/source/zendesk-chat/migrate-to-low-code` and still see the issue. Or we ask the user to reset?

Comment (bazarnov) [reply airbyte-integrations/connectors/source-zendesk-chat/source_zendesk_chat/manifest.yaml:269]: It's not the `ID` has changed, the actual change is in the `start_time`, but because the `id` was missing for this stream, as a PK, so i've added it to the schema. It was not my change or something, i've just transferred the logic from the old code base to the low-code.

File: airbyte-integrations/connectors/source-zendesk-chat/source_zendesk_chat/components/id_incremental_cursor.py
Comment (bazarnov) [reply airbyte-integrations/connectors/source-zendesk-chat/source_zendesk_chat/components/id_incremental_cursor.py:82]: I've covered this method with `unit_test`, so it's testable.

File: airbyte-integrations/connectors/source-zendesk-chat/source_zendesk_chat/components/bans_record_extractor.py
Comment (maxi297) [reply airbyte-integrations/connectors/source-zendesk-chat/source_zendesk_chat/components/bans_record_extractor.py:26]: And why would we have filtering?

Comment (bazarnov) [reply airbyte-integrations/connectors/source-zendesk-chat/source_zendesk_chat/components/bans_record_extractor.py:26]: It's not applied directly to the Zendesk situation, but i'd like to apply the best practices everywhere. If we have some sort of sorting in the underlying CDK (Low-code) methods - i think this could be removed. 

The answer to your question: 
> Why do we sort here? I see it was like that in the [previous implementation](https://github.com/airbytehq/airbyte/blob/0113c525548080f128258042d8663a415897ef8c/airbyte-integrations/connectors/source-zendesk-chat/source_zendesk_chat/streams.py#L271) but I'm curious if we could document why

My best guess is to just preserve the `ASC` order, after unnesting these 2 props: { "ip_address": {...}, "visitor": {...} }. For me it looks fine to have.

The regular case is when we have `semi-incremental` meaning we pull all the data from `start_date` but filterring out the records `<=` to `state`, if there is a cursor field suitable to track the STATE (updated, created, etc).

File: airbyte-integrations/connectors/source-zendesk-chat/source_zendesk_chat/manifest.yaml
Comment (bazarnov) [reply airbyte-integrations/connectors/source-zendesk-chat/source_zendesk_chat/manifest.yaml:269]: Here are the methods from Python Version of the `Source Zendesk Chat`:
- [Convertion to different format](https://github.com/airbytehq/airbyte/blob/1d6b93d6346de38f140fc1ef6b1a736902f4941c/airbyte-integrations/connectors/source-zendesk-chat/source_zendesk_chat/streams.py#L100-L107)
- [applying STATE level `cursor_field` format mutation](https://github.com/airbytehq/airbyte/blob/1d6b93d6346de38f140fc1ef6b1a736902f4941c/airbyte-integrations/connectors/source-zendesk-chat/source_zendesk_chat/streams.py#L123-L128)
- [applying RECORD level `cursor_field` format mutation](https://github.com/airbytehq/airbyte/blob/1d6b93d6346de38f140fc1ef6b1a736902f4941c/airbyte-integrations/connectors/source-zendesk-chat/source_zendesk_chat/streams.py#L149-L151)

These convert the date-time with microseconds to the RFC-3339 without microseconds.

1) Should I create another Custom Components to cover this and apply this as a separate record_extractor and make the streams to produce the: `"2020-12-14T04:08:32Z"` instead of `"2020-12-14T04:08:32.301292Z"`
2) Should I make the same with the STATE?

**Comment by bazarnov** [comment]
TODO:
Apply the meaningful message from this template : https://docs.google.com/document/d/1V3ybA24Lj_fy7IQwsVVzSxs46qGOsUydwlAQEYSQfcE/edit?pli=1

File: airbyte-integrations/connectors/source-zendesk-chat/source_zendesk_chat/components/bans_record_extractor.py
Comment (maxi297) [reply airbyte-integrations/connectors/source-zendesk-chat/source_zendesk_chat/components/bans_record_extractor.py:26]: Ok! All of this makes sense to me. Thanks for the detailed explanation! I'm good with resolving this comment. I'll let you decide if it is worth documenting this reasoning somewhere for the next dev that might jump on this one

File: airbyte-integrations/connectors/source-zendesk-chat/source_zendesk_chat/manifest.yaml
Comment (maxi297) [reply airbyte-integrations/connectors/source-zendesk-chat/source_zendesk_chat/manifest.yaml:269]: Ok! So we basically lose precision because of how we parse datetime, right? I honestly don't mind keeping the behavior of the low-code implementation. This is a breaking change though. I don't know what is best: we do it as part of this PR or in another issue? I'll let you decide what is best

**Review by maxi297 — APPROVED** [review]
Conditional approval given:
* Comment about id changing is resolved
* Comment about filtering/ordering is resolve
* Migration is properly documented

Thanks a lot @bazarnov !

**Review by maxi297 — APPROVED** [review]
Discrepancies I still see:
* [fine with this] accounts, goals, roles, shortcuts, skills and triggers: limit has been removed
* [fine with this] agent_timeline and chats: `start_time` query param is not the same but we've identify it only affects execution outside of docker
* [fine with this] bans: caused by a bug in the previous implementation

Hence, I confirm that this change is all good. Thanks @bazarnov !

**Pull request merged**
